docs/proposals/201912_thanos_binary_index_header.md

```diff
@@ -0,0 +1,191 @@
+
+This tradeoff is acceptable, due to total memory used for all operation should be much smaller. Additionally such query spanning all of the blocks and series are unlikely and should be blocked by simple `sample` limit.
+
+### Benchmark
+
+How to micro-benchmark such change? mmaping is outside of Go run time, which counts allocations etc.
+ 
+### Do we need to mmap?
+
+mmap adds a lot of complexity and confusion especially around monitoring it's memory usage as it does not appear on Go profiles.
+
+While mmap is great for random access against a big file, in fact the current implementation of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664)
+reallocates symbols, offsets and label name=value pairs while reading. This really defies the purpose of mmap as we want to combine all info in a dense few sequential sections of index-header binary format, 
+this file will be mostly read sequentially. Still, label values can be accessed randomly, that's why we propose to start with mmap straight away. 
+
+### Is initial long startup, really a problem?
+
+After initial startup with persistent disk, next startup should be quick due to cached files on disk for old blocks.
+Only new one will be iterated on. However still initial startup and adhoc syncs can be problematic.
+
+### LabelNames and LabelValues. 
+
+Currently all of those methods are getting labels values and label names across all blocks in the system.
+
+This will load **all** blocks to the system on every such call.
+
+We have couple of options:
+ 
+* Limit with time range: https://github.com/thanos-io/thanos/issues/1811
+* Cache e.g popular label values queries like `__name__`
+ 
+## Alternatives
+
+* Introduce faster JSON parser like [json-iterator/go](https://github.com/json-iterator/go)
+  * However, this does not contribute towards faster creation of such `index-header` from TSDB index. 
+* Design external memory compatible index instead of TSDB index.
+  * This is being discussed in another threads. While we should look on this ASAP, it's rather a long term plan.
+* Build `index-cache.json` on sidecar
+  * This unfortunately may require some extra memory for sidecar which we want to avoid.
+
+## Work Plan
+
+Step 1: Replace `index-cache.json` with `index-header`: 
+  * Implement building `index-header` file from pieces of TSDB index.
+  * Allow Store Gateway to build it instead of `index-cache.json` to disk.
+  * Allow Store Gateway to use `index-header` instead of `index-cache.json` for queries.
+  * Load those files using mmap
+  * Remove building `index-cache.json` in compactor.
+Step 2: Load on demand in query time:
+  * Allow Store Gateway to load `index-header` on demand from disk in query time.
+  * Add background job to unmmap/unload `index-header` for blocks that are unused (LRU).
```

Comment:
Not just unused but the least recently used blocks if the memory limit is violated, right?

Reply:
How we would know this?

Reply:
Sorry, I meant some kind of limit in terms of their size on disk because it would be really nice to avoid RAM usage blowing up in case of some queries. Having all of it in RAM is horrible, yes, but OTOH it gives the users a relatively certain thing: we know that at least that part is more or less equal over a period of time.